compiler/rustc_passes/src/dead.rs

```diff
@@ -72,24 +72,26 @@ fn adt_of<'tcx>(ty: &hir::Ty<'tcx>) -> Option<(LocalDefId, DefKind)> {
 }
 
 fn struct_all_fields_are_public(tcx: TyCtxt<'_>, id: LocalDefId) -> bool {
-    // treat PhantomData and positional ZST as public,
-    // we don't want to lint types which only have them,
-    // cause it's a common way to use such types to check things like well-formedness
-    tcx.adt_def(id).all_fields().all(|field| {
+    let adt_def = tcx.adt_def(id);
+
+    // skip types contain fields of unit and never type,
+    // it's usually intentional to make the type not constructible
+    let not_require_constructor = adt_def.all_fields().any(|field| {
         let field_type = tcx.type_of(field.did).instantiate_identity();
-        if field_type.is_phantom_data() {
-            return true;
-        }
-        let is_positional = field.name.as_str().starts_with(|c: char| c.is_ascii_digit());
-        if is_positional
-            && tcx
-                .layout_of(tcx.param_env(field.did).and(field_type))
-                .map_or(true, |layout| layout.is_zst())
-        {
-            return true;
-        }
-        field.vis.is_public()
-    })
+        field_type.is_unit() || field_type.is_never()
```

Comment:
```suggestion
        field_type.is_unit() || field_type.is_never() || field_type.is_phantom_data()
```
Could you merge the common logic for units, never types and phantom data?

Reply:
I thought only one field of `PhantomData`  does not imply that the whole type does not require a pub constructor, so the logic is a little different for `PhantomData`.

For example, if the whole type only have one field of `PhantomData<T>` and other fields of normal types, `PhantomData` is used to represent that the type owns `T`. So that marking types whose fields are all of `PhantomData` is more precise. In such cases, the whole type is usually used to check things like well-formedness.